porch/pkg/engine/engine.go

```diff
@@ -417,41 +422,91 @@ func (cad *cadEngine) ListFunctions(ctx context.Context, repositoryObj *configap
+	}).FetchResources(ctx, currUpstreamPkgRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package %s with ref %+v",
+			m.pkgName, *currUpstreamPkgRef)
+	}
 
-	// TODO: This is a hack
-	packageName := filepath.Base(m.task.Clone.Upstream.Git.Directory)
-	packageName = strings.TrimPrefix(packageName, ".git")
+	upstreamRevision, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchRevision(ctx, targetUpstream.UpstreamRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching revision for target upstream %s", targetUpstream.UpstreamRef.Name)
+	}
+	upstreamResources, err := (&PackageFetcher{
+		cad:               m.cad,
+		referenceResolver: m.referenceResolver,
+	}).FetchResources(ctx, targetUpstream.UpstreamRef, m.namespace)
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching resources for target upstream %s", targetUpstream.UpstreamRef.Name)
+	}
 
-	packageDir := filepath.Join(dir, packageName)
-	if err := kpt.PkgUpdate(ctx, ref, packageDir, kpt.PkgUpdateOpts{}); err != nil {
-		return repository.PackageResources{}, nil, err
+	klog.Infof("performing pkg upgrade operation for pkg %s resource counts local[%d] original[%d] upstream[%d]",
+		m.pkgName, len(resources.Contents), len(originalResources.Spec.Resources), len(upstreamResources.Spec.Resources))
+
+	// May be have packageUpdater part of engine to make it easy for testing ?
+	updatedResources, err := (&defaultPackageUpdater{}).Update(ctx,
+		resources,
+		repository.PackageResources{
+			Contents: originalResources.Spec.Resources,
+		},
+		repository.PackageResources{
+			Contents: upstreamResources.Spec.Resources,
+		})
+	if err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("error updating the package to revision %s", targetUpstream.UpstreamRef.Name)
 	}
 
-	loaded, err := loadResourcesFromDirectory(dir)
+	newUpstream, newUpstreamLock, err := upstreamRevision.GetUpstreamLock()
 	if err != nil {
-		return repository.PackageResources{}, nil, err
+		return repository.PackageResources{}, nil, fmt.Errorf("error fetching the resources for package revisions %s", targetUpstream.UpstreamRef.Name)
+	}
+	if err := kpt.UpdateKptfileUpstream("", updatedResources.Contents, newUpstream, newUpstreamLock); err != nil {
+		return repository.PackageResources{}, nil, fmt.Errorf("failed to apply upstream lock to package %q: %w", m.pkgName, err)
 	}
 
-	return loaded, m.task, nil
+	return updatedResources, m.task, nil
+}
+
+// Currently assumption is that downstream packages will be forked from a porch package.
+// As per current implementation, upstream package ref is stored in an old clone task, but this may
+// change so the logic of figuring out current upstream will live in this function.
+func (m *updatePackageMutation) currUpstream() (*api.PackageRevisionRef, error) {
```

Comment:
So this information would also be available in the `UpstreamLock` of the Kptfile in the package right? I'm thinking we might want to consider that the SoT, but we need to figure out how we want to handle the tasks.

Reply:
hmm..currently no. As per current implementation that @natasha41575 has in progress, the upstream info will be contain upstream Git(repo/dir/commit) or OCI(...) info etc, but it doesn't contain mapping to the porch packagerevision unless we store that as annotation in the `Kptfile` of the package. 

Or we do lookup of the packagerevision on the basis of upstream info in Kptfile.

Reply:
We could still look at the Kptfile though, right? We could look at the Kptfile in the `resources repository.PackageResources` that are  available in the `Apply` method above to analyze the Kptfile and extract the upstream information. (I'm assuming that those are the package resources.) We don't necessarily need to know the packagerevision to look at the Kptfile.

Reply:
> Although, I realize that the UpstreamRef that we're looking for might not be the same thing that is stored in the Kptfile's UpstreamLock.Ref.

That's is precisely the reason I am avoiding using the upstream info from the `Kptfile`. And as per our discussion today, I think it is reasonable to assume that every downstream package will have `clone` Task as the first task. So retrieving upstream info from `clone` task (like it is being done currently) seems to be a reasonable approach.